lib/vrl/stdlib/src/tag_types_externally.rs

```diff
@@ -0,0 +1,225 @@
+                result: Ok(r#"{ "integer": 123 }"#),
+            },
+            Example {
+                title: "object",
+                source: indoc! {r#"
+                    tag_types_externally({
+                        "message": "Hello world",
+                        "request": {
+                            "duration_ms": 67.9
+                        }
+                    })
+                "#},
+                result: Ok(
+                    r#"{ "message": { "bytes": "Hello world" }, "request": { "duration_ms": { "float": 67.9 } } }"#,
+                ),
+            },
+            Example {
+                title: "array",
+                source: r#"tag_types_externally(["foo", "bar"])"#,
+                result: Ok(r#"[{ "bytes": "foo" }, { "bytes": "bar" }]"#),
+            },
+        ]
+    }
+
+    fn compile(&self, mut arguments: ArgumentList) -> Compiled {
+        let value = arguments.required("value");
+
+        Ok(Box::new(TagTypesExternallyFn { value }))
+    }
+
+    fn parameters(&self) -> &'static [Parameter] {
+        &[Parameter {
+            keyword: "value",
+            kind: kind::ANY,
+            required: true,
+        }]
+    }
+}
+
+#[derive(Debug, Clone)]
+struct TagTypesExternallyFn {
+    value: Box<dyn Expression>,
+}
+
+impl Expression for TagTypesExternallyFn {
+    fn resolve(&self, ctx: &mut Context) -> Resolved {
+        let value = self.value.resolve(ctx)?;
+        let tagged_externally = tag_type_externally(value);
+
+        Ok(tagged_externally)
+    }
+
+    fn type_def(&self, state: &state::Compiler) -> TypeDef {
+        match self.value.type_def(state).kind() {
+            kind if kind.is_array() => TypeDef::new()
+                .infallible()
+                .array_mapped::<(), Kind>(map! { (): Kind::all() }),
+            _ => TypeDef::new()
+                .infallible()
+                .object::<(), Kind>(map! { (): Kind::all() }),
+        }
+    }
+}
+
+fn tag_type_externally(value: Value) -> Value {
+    let (key, value) = match value {
+        value @ Value::Bytes(_) => (Some("bytes"), value),
+        value @ Value::Integer(_) => (Some("integer"), value),
+        value @ Value::Float(_) => (Some("float"), value),
+        value @ Value::Boolean(_) => (Some("boolean"), value),
+        Value::Object(object) => (
+            None,
+            object
+                .into_iter()
+                .map(|(key, value)| (key, tag_type_externally(value)))
+                .collect::<BTreeMap<String, Value>>()
+                .into(),
+        ),
+        Value::Array(array) => (
+            None,
+            array
+                .into_iter()
+                .map(tag_type_externally)
+                .collect::<Vec<_>>()
+                .into(),
+        ),
+        value @ Value::Timestamp(_) => (Some("timestamp"), value),
+        value @ Value::Regex(_) => (Some("regex"), value),
+        Value::Null => (Some("null"), Value::Null),
```

Comment:
Likewise from [here](https://github.com/timberio/vector/issues/5727#issuecomment-774510068), nulls should just be removed..

Reply:
But in many cases `null` can carry information and would just be missing from the document then?

Or are you proposing to change it to

```suggestion
        Value::Null => (None, Value::Null),
```

Reply:
I have no strong opinions either way - just trying to interpret that comment.  Might be worth confirming with @binarylogic.

Reply:
I changed it to the more concise version now (meaning: `null` stays `null`).

My reasoning for keeping the `null`s is: If you want to remove them, you can do that independendly of this transformation. The reverse (adding them back if they are removed by default) would be quite hard.

Reply:
Agree with @pablosichert, it's fine to keep nulls. The purpose of this feature is to avoid type conflicts for storages that require it. I could see the presence of a field meaning something even if it does not have a value.